Update Celery by patching the imports

AmitPhulera opened the pull request:
## Product Description
<!-- For non-invisible changes, describe user-facing effects. -->
As a followup to the comment https://github.com/dimagi/commcare-hq/pull/31584#discussion_r868271327 
Since there were a lot of changes in the https://github.com/dimagi/commcare-hq/pull/31584, so we decided to patch the imports so that we can make the celery upgrade with minimal changes.
Most of the commits have been cherry-picked from #31584 
 
## Technical Summary
<!--
    Provide a link to the ticket or document which prompted this change,
    Describe the rationale and design decisions.
-->
[JIRA ticket](https://dimagi-dev.atlassian.net/browse/SAAS-13487)
## Feature Flag
<!-- If this is specific to a feature flag, which one? -->

## Safety Assurance

### Safety story
<!--
Describe how you became confident in this change, such as
local testing, why the change is inherently safe, and/or plans to limit the blast radius of a defect.
-->
The PR has been on staging for 2 months now and it has been thoroughly tested by QA.
### Automated test coverage

<!-- Identify the related test coverage and the tests it would catch -->

### QA Plan

<!--
- Describe QA plan that along with automated test coverages proves this PR is regression free
- Link to QA Ticket
-->
https://dimagi-dev.atlassian.net/browse/QA-4252

### Migrations
<!-- Delete this section if the PR does not contain any migrations -->
- [ ] The migrations in this code can be safely applied first independently of the code

<!-- Please link to any past code changes that are coordinated with this migration -->

### Rollback instructions

<!--
If this PR follows standards of revertability, check the box below.
Otherwise replace it with detailed instructions or reasons a rollback is impossible.
-->
There are some breaking migrations in this PR. To revert the PR the following steps should be followed.

1) Create a revert for this PR and wait for tests to pass.
2) Stop the celery service in the environment you are targeting.
   ```
   cchq <env> service celery stop
   ```
3) ssh into `django-manage` using
   ```
   cchq <env> ssh django_manage:0
   ```
4) Switch to cchq user
   ```
   sudo -iu cchq
   ```
5) Get to directly where HQ is installed and source the virutalenv
   ```
   cd /home/cchq/www/<env>/current && source python_env/bin/activate
   ```
6) Revert migrations to the previous versions
   ```
   ./manage.py migrate django_celery_results 0003
   ```
7) Merge the revert PR that was created in step 1.

8) Deploy to environment using
   ```
   cchq --control <env> deploy
   ```

### Labels & Review
- [x] Risk label is set correctly
- [x] The set of people pinged as reviewers is appropriate for the level of risk of the change

AmitPhulera committed:
update: revoking task logic

AmitPhulera committed:
fix: kombu.five issue

AmitPhulera committed:
update: warnings.py

AmitPhulera committed:
fix: tests

AmitPhulera committed:
add: corehq.apps.userreports.util to warninigs.py

AmitPhulera committed:
update base-requirements.in

AmitPhulera committed:
remove: celery references from requirement files and run make requirement

AmitPhulera committed:
add: decorator for @periodic_tasks

AmitPhulera committed:
update: celery app __init__.py

AmitPhulera committed:
migration.lock

AmitPhulera committed:
patch: celery import in manage.py

AmitPhulera committed:
fix-tests: remove timing context and query synclog db directly

AmitPhulera committed:
update: django-celery-result to point to github repo with the fix

AmitPhulera committed:
make requirements

AmitPhulera committed:
add: object patching to commcarehq_wsgi.py

AmitPhulera committed:
update: flower

AmitPhulera committed:
update:fix import path

millerdev commented on deployment/gunicorn/commcarehq_wsgi.py:
```diff
@@ -15,5 +15,8 @@
 patch_all(subprocess=True)
 patch_psycopg()
 
+from manage import run_patches
```
Can `run_patches` be imported above on line 3? (if it's needed at all, see below)

AmitPhulera commented on deployment/gunicorn/commcarehq_wsgi.py:
```diff
@@ -15,5 +15,8 @@
 patch_psycopg()
 
+from manage import run_patches
+run_patches()
```
Gunicorn was failing to start up until we have patched it in `commcarehq_wsgi.py`.
This caused staging to be inaccessible for some time.

millerdev replied:
Can the `run_patches()` in [gunicorn_conf.py](https://github.com/dimagi/commcare-hq/blob/b2fadfc628cca9a27e2791adf0d2666a26ff764b/deployment/gunicorn/gunicorn_conf.py#L17-L18) be removed since it was added here? I'm not sure if we've tested applying patches twice in the same process.

AmitPhulera committed:
update: django-celery-result location to dimagi owned repo

AmitPhulera committed:
move: import location

AmitPhulera committed:
bugfix: instantiate Task

AmitPhulera committed:
Merge remote-tracking branch 'origin/master' into ap/celery-upgrade-patch

AmitPhulera committed:
replace forked repo with original package

AmitPhulera committed:
remove the patch as it was done in commcarehq_wsgi.py

AmitPhulera committed:
update patch

AmitPhulera committed:
add: a wrapper over shared_task decorator

AmitPhulera committed:
update: default celery serailizer to pickle

millerdev commented on corehq/apps/celery/shared_task.py:
```diff
@@ -0,0 +1,25 @@
+    """
+    kwargs['serializer'] = kwargs.setdefault('serializer', 'json')
+
+    def wrapper(fn):
+        return shared_task(*args, **kwargs)(fn)
+
+    return wrapper
```
nit: it can be nice for debugging to have the function name be something more useful than `wrapper`.
```suggestion
    def task(fn):
        return shared_task(*args, **kwargs)(fn)

    return task
```

millerdev commented on settings.py:
```diff
@@ -584,7 +584,9 @@
-CELERY_TASK_SERIALIZER = 'json'  # Default value in celery 4.x
+# Default serializer should be changed back to 'json' after
+# https://github.com/celery/celery/issues/6759 is fixed
+CELERY_TASK_SERIALIZER = 'pickle'
```
An inline comment will make this more visible when searching with grep
```suggestion
CELERY_TASK_SERIALIZER = 'pickle'  # default is 'json'. See corehq.apps.celery.shared_task
```

millerdev reviewed (commented):
Looking good. Few nitpicks and suggestions.

AmitPhulera committed:
Apply suggestions from code review

Co-authored-by: Daniel Miller <dmiller@dimagi.com>

AmitPhulera committed:
add: support for task decorater with no args

millerdev commented on corehq/apps/celery/shared_task.py:
```diff
@@ -22,6 +22,9 @@ def task(*args, **kwargs):
     """
     kwargs.setdefault('serializer', 'json')
 
+    if len(args) == 1 and callable(args[0]):
+        return shared_task(**kwargs)(args[0])
```
The commit message "add: support for task decorater with no args" implies that `kwargs` must be empty.

I can imagine an admittedly weird, possibly invalid, scenario that would trigger this branch, but probably should not:

```py
@task(dont_decorate_this_callable, max_retries=3)
def do_something():
    ...
```

Suggested change:
```suggestion
    if len(args) == 1 and callable(args[0]) and not kwargs:
        return shared_task()(args[0])
```

AmitPhulera replied:
I don't think this should be an issue but I like the idea of safeguarding it as much as possible.

AmitPhulera committed:
ensure that kwargs should be empty

Co-authored-by: Daniel Miller <dmiller@dimagi.com>

AmitPhulera committed:
move the args check logic before setting serializer

millerdev commented on corehq/apps/celery/shared_task.py:
```diff
@@ -0,0 +1,32 @@
+    """
+    
+    if len(args) == 1 and callable(args[0]) and not kwargs:
+        return shared_task()(args[0])
```
Should this specify the serializer? Sorry, I missed that in my previous suggestion.
```suggestion
        return shared_task(serializer='json')(args[0])
```

AmitPhulera replied:
Yep. It should. Thanks for pointing it out. Earlier we were setting serializer before but after https://github.com/dimagi/commcare-hq/pull/31713/commits/506bfd326851f1436ed6c1646c3ef4fd93344bdb we need to set it.

AmitPhulera committed:
add serializer while returning shared_task

Co-authored-by: Daniel Miller <dmiller@dimagi.com>

AmitPhulera committed:
lint

calellowitz commented on settings.py:
```diff
@@ -584,7 +584,9 @@
-CELERY_TASK_SERIALIZER = 'json'  # Default value in celery 4.x
+# Default serializer should be changed back to 'json' after
+# https://github.com/celery/celery/issues/6759 is fixed
+CELERY_TASK_SERIALIZER = 'pickle'  # default is 'json'. See corehq.apps.celery.shared_task
```
Can you make this comment a bit clearer. Something like "this value in commcare hq code". I read this comment to say "celery defaults to json, but we are setting this value to default to pickle in commcare hq," which isn't true. We still default to json, and I think that should be much clearer without someone needing to figure out how the patch we made works.

AmitPhulera replied:
https://github.com/dimagi/commcare-hq/pull/31713/commits/ede3093687278c3119d0461e32a5b586513c3b24

AmitPhulera committed:
Merge remote-tracking branch 'origin/master' into ap/celery-upgrade-patch

AmitPhulera committed:
update: comment text

AmitPhulera committed:
update: docstring

AmitPhulera committed:
add: namespacing while loading config from celery settings

While testing out management commands I realised that they are timing out, on looking I figured that we need to initialize app config with the namesapce

AmitPhulera committed:
remove hack which is not required in newer versions

AmitPhulera committed:
Merge remote-tracking branch 'origin/master' into ap/celery-upgrade-patch

AmitPhulera committed:
keep queue available in options so that

1) Periodic tasks can be called with delay too.\n 2) errors in celery_monitoring go away

AmitPhulera committed:
Merge remote-tracking branch 'origin/master' into ap/celery-upgrade-patch

AmitPhulera committed:
set serializer as json for periodic tasks too

AmitPhulera committed:
set task_queue to NA when it is not there in the task

specifically happening for send_unknown_user_type_stats tasks where intermittently task_queue is not set for the task

AmitPhulera committed:
pass default queue for periodic tasks

AmitPhulera committed:
remove N/A check for queue name

AmitPhulera committed:
set default queue in shared_task if not queue is defined

AmitPhulera committed:
set default queue for periodic tasks as celery

calellowitz reviewed (approved):
It looks like a lot of logic is replicated between the new `task` function and the `periodic_task` one. Is it possible to share that logic instead so we only have to maintain it in one place? It might be possible to have this done by having `periodic_task` call the `task` method, although I am not sure how those interact specifically.

AmitPhulera committed:
let shared_task handle the periodic_task config settings

AmitPhulera committed:
update: periodic_task to use the hq task

AmitPhulera commented:
@calellowitz  See  8735ff206770f10dde1e2bec543a04983480d75a and ee609333c5717e3a8bfd74656f00ecef61a5257d

calellowitz commented on corehq/apps/celery/shared_task.py:
```diff
@@ -28,6 +30,10 @@ def task(*args, **kwargs):
+    kwargs.setdefault('options', {})
+
+    if kwargs.get('base') == PeriodicTask:
+        kwargs['options']['queue'] = kwargs.get('queue')
```
Why do periodic tasks use ` kwargs['options']['queue']` vs ` kwargs['queue']` like shared task?

AmitPhulera replied:
Ohh! Beat uses `kwargs['options']['queue']` in order to determine which queue would Periodic tasks should run on.
But for shared tasks `kwargs['queue']` would work as celery is directly running them.

calellowitz reviewed (approved):
one question but looks good
